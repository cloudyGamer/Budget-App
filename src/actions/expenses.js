 /* 
  * To change this license header, choose License Headers in Project Properties.
  * To change this template file, choose Tools | Templates
  * and open the template in the editor.
  */
 import uuid from 'uuid';
 import database from '../firebase/firebase';
 //ADD EXPENSE
 //pay attention - this fnc takes a parameter; an object, it is using destructuring and default values. Notice
 //it also is implicitly returning an object.
 //we have already prepared the add expense object in startAddExpense
 //const uid = "zehAI6q4h6VPkZUouJ32e3Pd7Lt1";
 export const addExpense = (expense) => ({
           type: 'ADD_EXPENSE',
           expense

      });
//look; here we are taking in expense data and again destructuring it, we are creating consts here .
 export const startAddExpense = (expenseData = {}) => {
      return (dispatch, getState) => {
           const uid = getState().auth.uid;
           const {
                description = '',
                amount = 0,
                note = '',
                createdAt = 0
           } = expenseData;
//these consts are then added to an object to be sent to databse
           const expense = {description, amount, note, createdAt};
           return database.ref(`users/${uid}/expenses`).push(expense)
             //now when that's been done successfully we will add the same data to Redux
             .then((ref) => {
                  //note that firebase has already given us an id and we can access that ref
                  //we create a new obj to hold id and expense data
                  dispatch(addExpense({
                       id: ref.key,
                       ...expense
                  }));
             });
      };
 };

 export const removeExpense = ({ id } = {}) => ({
           type: 'REMOVE_EXPENSE',
           id
      });

 export const startRemoveExpense = ({ id } = {}) => {
      return (dispatch, getState) => {
           const uid = getState().auth.uid;
           return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
                dispatch(removeExpense({id}));
           });
      };
 };

 export const editExpense = (id, updates) => ({
           type: 'EDIT_EXPENSE',
           id,
           updates

      });

 export const startEditExpense = (id, updates) => {
      return(dispatch, getState) => {
           const uid = getState().auth.uid;
           return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(
             () => {
                dispatch(editExpense(id, updates));
           }
           );
      };
 };

//set expenses
 export const setExpenses = (expenses) => ({
           type: 'SET_EXPENSES',
           expenses
      });

 export const startSetExpenses = () => {
      //expenses are already there when this thing runs
      //this was your major source of confusion
      //seems like nothing there just randomly adds bits to the database automatically
      //but the big gain is that you can refresh the browser and the data persists
      //this is what he means by set up expenses
      return (dispatch, getState) => {
           //return reference 1
           const uid = getState().auth.uid;
           return database.ref(`users/${uid}/expenses`).once('value').then(
             (snapshot) => {
                const expensesData = [];
                const val = snapshot.forEach((snap) => {
                     expensesData.push({
                          id: snap.key,
                          ...snap.val()
                     });
                });
                //dispatch
                dispatch(setExpenses(expensesData));
           }
           );
      };
 };

