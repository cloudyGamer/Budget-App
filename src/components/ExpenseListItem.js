
import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
 //description //createdAt //amount
 //use map - the map goes in expense list. pass the individual items as props
// load a locale
numeral.register('locale', 'irl', {
    delimiters: {
        thousands: ' ',
        decimal: ','
    },
    abbreviations: {
        thousand: 'k',
        million: 'm',
        billion: 'b',
        trillion: 't'
    },
    ordinal : function (number) {
        return number === 1 ? 'st' : 'th';
    },
    currency: {
        symbol: 'â‚¬'
    }
});

// switch between locales
numeral.locale('irl');
export const ExpenseListItem = ({dispatch, description, amount, createdAt, id}) => (
  <div xmlns="http://www.w3.org/1999/xhtml">
  <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
    <p>
    {numeral(amount/100).format('$0,0.00')} 
    - 
    {moment(createdAt).format('MMMM Do, YYYY')}</p>

  </div>  

  );

export default ExpenseListItem;
