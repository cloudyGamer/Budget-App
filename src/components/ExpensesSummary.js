import React from 'react';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import { connect } from 'react-redux';

//export const ExpensesSummary = (props) => (
//   {/*  <div>
//   {
//   props.expenses.length !== 0 &&
//     <div>
//   <a>Displaying</a>  <a>{props.expenses.length}</a> { props.expenses.length === 1 ? (<a>expense</a> ) : (<a>expenses</a>)}
//      <a> with a value of </a> 
//       <a>{props.expenses
//       .map((expense)=>expense.amount)
//       .reduce((total,num) => total+num, 0)}
//       </a>
//     </div> 
//          
//          }
//   </div>*/}  
//
//   );

     
export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
     const expenseWord = expenseCount ===1 ? 'expense':'expenses';
     const formattedTotal = numeral(expensesTotal/100).format('â‚¬0,0.00');
     return (
       <div>
       <h1>Viewing {expenseCount} {expenseWord} totalling {formattedTotal} euro</h1>
       </div>  
       );
};  

const mapStateToProps = (state) => {
     const visibleExpenses = selectExpenses(state.expenses, state.filters);
      return {
           expenseCount: visibleExpenses.length,
           expensesTotal: selectExpensesTotal(visibleExpenses)
           
      };
 };

export default connect(mapStateToProps)(ExpensesSummary);
