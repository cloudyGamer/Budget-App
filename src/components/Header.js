 /* 
  * To change this license header, choose License Headers in Project Properties.
  * To change this template file, choose Tools | Templates
  * and open the template in the editor.
  */
 
import React from 'react'; 
 import {Link,NavLink} from 'react-router-dom';
const Header = () => (
  <header>
  <h1>Expensify App</h1>
  <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
  <NavLink to="/create" activeClassName="is-active">Add Expense</NavLink>
  <NavLink to="/help" activeClassName="is-active">Help</NavLink>
  </header>
  );

 export default Header;