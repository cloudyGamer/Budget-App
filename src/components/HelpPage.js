 /* 
  * To change this license header, choose License Headers in Project Properties.
  * To change this template file, choose Tools | Templates
  * and open the template in the editor.
  */
  
import React from 'react'; 
 const HelpPage = () => (
  <div>
  This is from my HelpPage  component
  </div>  

  );
 
 export default HelpPage;