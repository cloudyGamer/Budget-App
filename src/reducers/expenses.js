 /* 
  * To change this license header, choose License Headers in Project Properties.
  * To change this template file, choose Tools | Templates
  * and open the template in the editor.
  */
 
 const expensesReducerDefaultState = [];
 export default   (state = expensesReducerDefaultState,action) => {
      switch(action.type){
          case 'ADD_EXPENSE':
          return [
                  ...state,
                  action.expense
                 ];
           
          case 'REMOVE_EXPENSE':
          return state.filter(({id}) => id !== action.id);
          
          case 'EDIT_EXPENSE':
          return state.map((expense)=>{
              if(expense.id===action.id){
                   return{
                         ...expense,
                         ...action.updates
                   };
              }else{
                   return expense;
              }
         }
           );
          
           default:
                  return state;
      }
};

