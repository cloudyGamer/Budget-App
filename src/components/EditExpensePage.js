 
import React from 'react'; 
import { connect } from 'react-redux';
import { startEditExpense,editExpense,removeExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
     onSubmit = (expense) => {
          //console.log("updated",expense);
          this.props.startEditExpense(this.props.expense.id,expense);
          this.props.history.push(`/`);
     };
     
     onRemove = () => {
          this.props.startRemoveExpense({id: this.props.expense.id});
          this.props.history.push(`/`);
     };
     render(){
          return(
               <div>
                 <h1>Expense Form</h1>    
                    <ExpenseForm xmlns="http://www.w3.org/1999/xhtml"
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    /> 
                    <button 
                    onClick={this.onRemove}>Remove</button>
               </div>  
            );
     }
};
const mapDispatchToProps = (dispatch,props) => ({
     startEditExpense: (expenseid,expense) => dispatch(startEditExpense(expenseid,expense)),
     startRemoveExpense: (expense) => dispatch(startRemoveExpense(expense))
});

const mapStateToProps = (state,props) => {
     return {
          expense : state.expenses.find((expense) => expense.id===props.match.params.id)
     };
};

 export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);