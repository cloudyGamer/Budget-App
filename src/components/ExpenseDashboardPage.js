 /* 
  * To change this license header, choose License Headers in Project Properties.
  * To change this template file, choose Tools | Templates
  * and open the template in the editor.
  */

import React from 'react'; 
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
export const ExpenseDashboardPage = () => (
  <div xmlns="http://www.w3.org/1999/xhtml">
     <ExpensesSummary />
     <ExpenseListFilters />
     <ExpenseList />
  </div>  

  );

 export default ExpenseDashboardPage;