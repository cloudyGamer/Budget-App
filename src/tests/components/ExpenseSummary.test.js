import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpenses from '../../selectors/expenses';
import expenses from '../fixtures/expenses';
//let wrapper;
//beforeAll(
//  
//  );

test('should render ExpensesSummary with 1 expense', () => {
     const wrapper= shallow(<ExpensesSummary xmlns="http://www.w3.org/1999/xhtml" expenseCount={1} expensesTotal={235}/>);
     expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary with multiple expenses', () => {
     const wrapper= shallow(<ExpensesSummary xmlns="http://www.w3.org/1999/xhtml" expenseCount={3} expensesTotal={235455}/>);
     expect(wrapper).toMatchSnapshot();
});