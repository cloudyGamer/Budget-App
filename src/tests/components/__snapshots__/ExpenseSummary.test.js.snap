// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    2,354.55
     euro
  </h1>
</div>
`;

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    2.35
     euro
  </h1>
</div>
`;
